feat: Auto reconcile after new draft voucher submission in reco tool

marination opened the pull request:
- On "Edit in Full Page", create a new voucher in a new tab and listen to new voucher doc updates
- If voucher is deleted after creation, turn off listener
- If simply updated (saved), do nothing
- If submitted, reconcile with current transaction and update UI
- **If user moves to next transaction**, turn off listener. We don't want random transactions getting reconciled

> These actions will take place on a transaction given that it is the active selection.

Magic 🪄
![2023-09-11 20 45 57](https://github.com/alyf-de/banking/assets/25857446/5ddc211d-1241-4862-a21c-55f2273911d9)

marination committed:
feat: Auto reconcile after new draft voucher submission in reco tool

- On "Edit in Full Page", listen to new voucher doc updates
- If voucher is deleted, turn off listener
- If updated, do nothing
- If submitted, reconcile with current transaction and update UI
- If user moves to next transaction without taking action, turn off listener

marination commented:
To fix:

- [ ] Check perms in APIs (separate PR)
- [x] Add type hints
- [x] Make sure API return values are predictable

barredterra committed:
Merge branch 'version-14' into edit-in-full-page

barredterra commented on banking/klarna_kosma_integration/doctype/bank_reconciliation_tool_beta/bank_reconciliation_tool_beta.py:
```diff
@@ -215,16 +207,33 @@ def create_payment_entry_bts(
+
+
+@frappe.whitelist()
+def reconcile_entry(transaction_name, amount, voucher_type, voucher_name):
```
```suggestion
def reconcile_voucher(transaction_name, amount, voucher_type, voucher_name):
```

Let's make the function name match the parameter names?

barredterra commented on banking/public/js/bank_reconciliation_beta/actions_panel/create_tab.js:
```diff
@@ -32,10 +32,23 @@ erpnext.accounts.bank_reconciliation.CreateTab = class CreateTab {
 	}
 
 	edit_in_full_page() {
+		var me = this;
```
IIRC, `me === this` inside an arrow function.

barredterra commented on banking/public/js/bank_reconciliation_beta/actions_panel/create_tab.js:
```diff
@@ -89,6 +102,41 @@ erpnext.accounts.bank_reconciliation.CreateTab = class CreateTab {
+					);
+				}
+			}
+		})
```
```suggestion
		});
```

barredterra commented:
Functionality LGTM

marination committed:
fix: Type Hints and misc

- Add type hints to all whitelisted functions in Bank Reco Beta
- Rename `reconcile_entry` to `reconcile_voucher`
- Make `reconcile_voucher` return types predictable. It returns dicts or BT doc (all interpreted as object in JS). Tweak JS checks.

marination committed:
Merge branch 'version-14' into edit-in-full-page
